src/tallies/tally_filter_sph_harm.F90

```diff
@@ -0,0 +1,241 @@
+module tally_filter_sph_harm
+
+  use, intrinsic :: ISO_C_BINDING
+
+  use hdf5, only: HID_T
+
+  use constants
+  use error
+  use hdf5_interface
+  use math,                only: calc_pn, calc_rn
+  use particle_header,     only: Particle
+  use string,              only: to_str, to_lower, to_f_string
+  use tally_filter_header
+  use xml_interface
+
+  implicit none
+  private
+  public :: openmc_sphharm_filter_get_order
+  public :: openmc_sphharm_filter_get_cosine
+  public :: openmc_sphharm_filter_set_order
+  public :: openmc_sphharm_filter_set_cosine
+
+  integer, parameter :: COSINE_SCATTER = 1
+  integer, parameter :: COSINE_PARTICLE = 2
+
+!===============================================================================
+! LEGENDREFILTER gives Legendre moments of the change in scattering angle
+!===============================================================================
+
+  type, public, extends(TallyFilter) :: SphericalHarmonicsFilter
+    integer :: order
+    integer :: cosine
+  contains
+    procedure :: from_xml
+    procedure :: get_all_bins
+    procedure :: to_statepoint
+    procedure :: text_label
+  end type SphericalHarmonicsFilter
+
+contains
+
+!===============================================================================
+! SphericalHarmonicsFilter methods
+!===============================================================================
+
+  subroutine from_xml(this, node)
+    class(SphericalHarmonicsFilter), intent(inout) :: this
+    type(XMLNode), intent(in) :: node
+
+    character(MAX_WORD_LEN) :: temp_str
+
+    ! Get specified order
+    call get_node_value(node, "order", this % order)
+    this % n_bins = (this % order + 1)**2
+
+    ! Determine how cosine term is to be treated
+    if (check_for_node(node, "cosine")) then
+      call get_node_value(node, "cosine", temp_str)
+      select case (to_lower(temp_str))
+      case ('scatter')
+        this % cosine = COSINE_SCATTER
+      case ('particle')
+        this % cosine = COSINE_PARTICLE
+      end select
+    else
+      this % cosine = COSINE_PARTICLE
+    end if
+  end subroutine from_xml
+
+  subroutine get_all_bins(this, p, estimator, match)
+    class(SphericalHarmonicsFilter), intent(in)  :: this
+    type(Particle),      intent(in)  :: p
+    integer,             intent(in)  :: estimator
+    type(TallyFilterMatch),   intent(inout) :: match
+
+    integer :: i, j, n
+    integer :: num_nm
+    real(8) :: wgt
+    real(8) :: rn(2*this % order + 1)
+
+    ! TODO: Use recursive formula to calculate higher orders
+    j = 0
+    do n = 0, this % order
+      ! Determine cosine term for scatter expansion if necessary
+      if (this % cosine == COSINE_SCATTER) then
+        wgt = calc_pn(n, p % mu)
+      else
+        wgt = ONE
+      end if
+
+      ! Calculate n-th order spherical harmonics for (u,v,w)
+      num_nm = 2*n + 1
+      rn(1:num_nm) = calc_rn(n, p % last_uvw)
```

Comment:
Should be uvw for flux and last_uvw for scatter ? Can't redo the math right now but it used to be like this in tally.f90

Reply:
Yes, there were two different cases before in tally.F90. However, there was a change introduced in #910 such that last_uvw and uvw will actually be the same, so there's no need for us to have separate cases now.